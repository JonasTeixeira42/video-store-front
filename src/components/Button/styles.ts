import styled, { css, DefaultTheme } from 'styled-components';
import { darken } from 'polished';

import { BackgroundColors, ButtonProps } from '.';

type WrapperProps = {
  hasIcon: boolean;
  hasChildren: boolean;
} & Pick<ButtonProps, 'size' | 'color' | 'background'>;

const wrapperModifiers = {
  xsmall: (theme: DefaultTheme) => css`
    p {
      padding: ${theme.spacings.xxsmall};
    }
  `,

  small: (theme: DefaultTheme) => css`
    p {
      padding: ${theme.spacings.xsmall};
    }
  `,

  large: (theme: DefaultTheme) => css`
    p {
      padding: ${theme.spacings.xsmall} ${theme.spacings.small};
    }
  `,

  withIcon: (theme: DefaultTheme, backgroundColor: BackgroundColors) => css`
    background: ${darken(0.1, theme.colors[backgroundColor])};

    &:hover {
      background: ${darken(0.2, theme.colors[backgroundColor])};

      & > div {
        background: ${darken(0.1, theme.colors[backgroundColor])};
      }
    }
  `,
};

export const Wrapper = styled.button<WrapperProps>`
  ${({ theme, size, hasIcon, color, background, hasChildren }) => css`
    display: flex;
    flex-direction: row;
    align-items: center;
    justify-content: center;

    font-weight: 600;
    font-size: ${theme.font.sizes.medium};
    border-radius: ${theme.border.radius};
    border: 0;
    outline: 0;
    background: ${theme.colors[background!]};
    color: ${theme.colors[color!]};
    cursor: pointer;
    transition: background 0.3s;

    &:hover {
      background: ${darken(0.1, theme.colors[background!])};
    }

    div {
      display: flex;
      padding: ${theme.spacings.xsmall};
      background: ${theme.colors[background!]};
      border-radius: 0 ${theme.border.radius} ${theme.border.radius} 0;
      margin: 0 auto;
      transition: background 0.3s;
    }

    ${!!size && wrapperModifiers[size!](theme)};

    ${hasIcon && wrapperModifiers.withIcon(theme, background!)}

    ${!hasChildren &&
    css`
      div {
        border-radius: ${theme.border.radius};
      }
    `}
  `}
`;
