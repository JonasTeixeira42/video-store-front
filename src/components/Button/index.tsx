import React, { ButtonHTMLAttributes } from 'react';

import * as S from './styles';

type ButtonType = ButtonHTMLAttributes<HTMLButtonElement>;

export type BackgroundColors = 'white' | 'primary' | 'secondary';

export type ButtonProps = {
  size?: 'small' | 'large';
  color?: 'white' | 'lightBlack';
  background?: BackgroundColors;
  icon?: JSX.Element;
} & ButtonType;

const Button = ({
  children,
  icon,
  size = 'large',
  color = 'white',
  background = 'secondary',
  ...props
}: ButtonProps) => {
  return (
    <S.Wrapper
      size={size}
      hasIcon={!!icon}
      hasChildren={!!children}
      color={color}
      background={background}
      {...props}
    >
      {!!children && <p>{children}</p>}

      {!!icon && <div>{icon}</div>}
    </S.Wrapper>
  );
};

export default Button;
