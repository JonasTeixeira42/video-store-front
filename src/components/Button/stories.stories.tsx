import { Story, Meta } from '@storybook/react/types-6-0';
import { AddShoppingCart } from '@styled-icons/material-outlined/AddShoppingCart';

import Button, { ButtonProps } from '.';

export default {
  title: 'Button',
  component: Button,
  argTypes: {
    children: {
      type: 'string',
    },
    icon: {
      type: '',
    },
  },
} as Meta;

export const Default: Story<ButtonProps> = (args) => <Button {...args} />;

Default.args = {
  children: 'Add Movies',
  icon: <AddShoppingCart size={24} />,
};

export const WithIcon: Story<ButtonProps> = (args) => <Button {...args} />;

WithIcon.args = {
  size: 'small',
  icon: <AddShoppingCart size={24} />,
};
