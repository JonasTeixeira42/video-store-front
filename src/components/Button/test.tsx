import { screen } from '@testing-library/react';
import { renderWithTheme } from 'utils/tests/helpers';
import { AddBox } from '@styled-icons/material-outlined/';

import Button from '.';

describe('<Button />', () => {
  it('should render the button with large size', () => {
    const { container } = renderWithTheme(<Button>Movies</Button>);

    const button = screen.getByRole('button', { name: /movies/i });

    expect(button).toBeInTheDocument();

    expect(button.firstElementChild).toHaveStyle({
      padding: '1.6rem 2.4rem',
    });

    expect(button).toHaveStyle({
      color: '#FAFAFA',
    });

    expect(container.firstChild).toMatchSnapshot();
  });

  it('should render small size when size passed', () => {
    renderWithTheme(<Button size="small">Movies</Button>);

    expect(
      screen.getByRole('button', { name: /movies/i }).firstElementChild,
    ).toHaveStyle({
      padding: '1.6rem',
    });
  });

  it('should render black color when color passed', () => {
    renderWithTheme(<Button color="lightBlack">Movies</Button>);

    expect(screen.getByRole('button', { name: /movies/i })).toHaveStyle({
      color: '#3D3D4D',
    });
  });

  it('should render icon version', () => {
    renderWithTheme(<Button icon={<AddBox data-testid="icon" />} />);

    expect(screen.getByTestId('icon').parentElement).toBeInTheDocument();
  });

  it('should render secondary color background by default', () => {
    renderWithTheme(<Button icon={<AddBox data-testid="icon" />} />);

    expect(screen.getByTestId('icon').parentElement).toHaveStyle({
      background: '#04d361',
    });
  });

  it('should render primary color background when background passed', () => {
    renderWithTheme(
      <Button background="primary" icon={<AddBox data-testid="icon" />} />,
    );

    expect(screen.getByTestId('icon').parentElement).toHaveStyle({
      background: '#F231A5',
    });
  });

  it('should render white color background when background passed', () => {
    renderWithTheme(
      <Button background="white" icon={<AddBox data-testid="icon" />} />,
    );

    expect(screen.getByTestId('icon').parentElement).toHaveStyle({
      background: '#FAFAFA',
    });
  });
});
