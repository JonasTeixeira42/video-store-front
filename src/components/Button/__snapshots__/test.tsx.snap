// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Button /> should render the button with large size 1`] = `
.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  font-weight: 600;
  font-size: 1.6rem;
  border-radius: 0.8rem;
  border: 0;
  outline: 0;
  background: #04d361;
  color: #FAFAFA;
  cursor: pointer;
  -webkit-transition: background 0.3s;
  transition: background 0.3s;
}

.c0:hover {
  background: #03a14a;
}

.c0 div {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  padding: 1.6rem;
  background: #04d361;
  border-radius: 0 0.8rem 0.8rem 0;
  margin: 0 auto;
  -webkit-transition: background 0.3s;
  transition: background 0.3s;
}

.c0 p {
  padding: 1.6rem 2.4rem;
}

<button
  class="c0"
  color="white"
>
  <p>
    Movies
  </p>
</button>
`;
